File: code/packages/csharp/irc-proto/IrcProto.cs
using System.Text;

namespace CodingAdventures.IrcProto;

/// <summary>
/// A single parsed IRC protocol message.
/// </summary>
public sealed record Message(string? Prefix, string Command, IReadOnlyList<string> Params);

/// <summary>
/// Raised when a raw line cannot be parsed as an IRC message.
/// </summary>
public sealed class ParseError : Exception
{
    /// <summary>Create a parse error with a human-readable message.</summary>
    public ParseError(string message) : base(message)
    {
    }
}

/// <summary>
/// Pure IRC message parsing and serialization helpers.
/// </summary>
public static class IrcProto
{
    /// <summary>The package version.</summary>
    public const string Version = "0.1.0";

    private const int MaxParams = 15;

    /// <summary>Parse one IRC line with any trailing CRLF already stripped.</summary>
    public static Message Parse(string line)
    {
        ArgumentNullException.ThrowIfNull(line);
        if (line.Length == 0 || string.IsNullOrWhiteSpace(line))
        {
            throw new ParseError($"empty or whitespace-only line: {line}");
        }

        var rest = line;
        string? prefix = null;
        if (rest.StartsWith(':'))
        {
            var spacePosition = rest.IndexOf(' ', StringComparison.Ordinal);
            if (spacePosition == -1)
            {
                throw new ParseError($"line has prefix but no command: {line}");
            }

            prefix = rest[1..spacePosition];
            rest = rest[(spacePosition + 1)..];
        }

        var commandEnd = rest.IndexOf(' ', StringComparison.Ordinal);
        string commandRaw;
        if (commandEnd == -1)
        {
            commandRaw = rest;
            rest = string.Empty;
        }
        else
        {
            commandRaw = rest[..commandEnd];
            rest = rest[(commandEnd + 1)..];
        }

        var command = commandRaw.ToUpperInvariant();
        if (command.Length == 0)
        {
            throw new ParseError($"could not extract command from line: {line}");
        }

        var parameters = new List<string>();
        while (rest.Length > 0)
        {
            if (rest.StartsWith(':'))
            {
                parameters.Add(rest[1..]);
                break;
            }

            var spacePosition = rest.IndexOf(' ', StringComparison.Ordinal);
            if (spacePosition == -1)
            {
                parameters.Add(rest);
                break;
            }

            parameters.Add(rest[..spacePosition]);
            rest = rest[(spacePosition + 1)..];

            if (parameters.Count == MaxParams)
            {
                break;
            }
        }

        return new Message(prefix, command, parameters.AsReadOnly());
    }

    /// <summary>Serialize a message to CRLF-terminated UTF-8 IRC wire bytes.</summary>
    public static byte[] Serialize(Message message)
    {
        ArgumentNullException.ThrowIfNull(message);
        var parts = new List<string>();
        if (message.Prefix is not null)
        {
            parts.Add($":{message.Prefix}");
        }

        parts.Add(message.Command);
        for (var i = 0; i < message.Params.Count; i++)
        {
            var parameter = message.Params[i];
            var isLast = i == message.Params.Count - 1;
            parts.Add(isLast && parameter.Contains(' ') ? $":{parameter}" : parameter);
        }

        return Encoding.UTF8.GetBytes(string.Join(" ", parts) + "\r\n");
    }
}

File: code/packages/fsharp/irc-proto/IrcProto.fs
namespace CodingAdventures.IrcProto.FSharp

open System
open System.Text

type Message =
    { Prefix: string option
      Command: string
      Params: string list }

exception ParseError of string

[<RequireQualifiedAccess>]
module IrcProto =
    [<Literal>]
    let Version = "0.1.0"

    let private maxParams = 15

    let parse (line: string) =
        if isNull line then
            nullArg "line"

        if line.Length = 0 || String.IsNullOrWhiteSpace line then
            raise (ParseError $"empty or whitespace-only line: {line}")

        let mutable rest = line

        let prefix =
            if rest.StartsWith ":" then
                let spacePosition = rest.IndexOf ' '

                if spacePosition = -1 then
                    raise (ParseError $"line has prefix but no command: {line}")

                let value = rest.Substring(1, spacePosition - 1)
                rest <- rest.Substring(spacePosition + 1)
                Some value
            else
                None

        let commandRaw =
            let commandEnd = rest.IndexOf ' '

            if commandEnd = -1 then
                let value = rest
                rest <- ""
                value
            else
                let value = rest.Substring(0, commandEnd)
                rest <- rest.Substring(commandEnd + 1)
                value

        let command = commandRaw.ToUpperInvariant()

        if command.Length = 0 then
            raise (ParseError $"could not extract command from line: {line}")

        let parameters = ResizeArray<string>()

        while rest.Length > 0 do
            if rest.StartsWith ":" then
                parameters.Add(rest.Substring 1)
                rest <- ""
            else
                let spacePosition = rest.IndexOf ' '

                if spacePosition = -1 then
                    parameters.Add rest
                    rest <- ""
                else
                    parameters.Add(rest.Substring(0, spacePosition))
                    rest <- rest.Substring(spacePosition + 1)

                    if parameters.Count = maxParams then
                        rest <- ""

        { Prefix = prefix
          Command = command
          Params = parameters |> Seq.toList }

    let serialize (message: Message) =
        let parts = ResizeArray<string>()

        match message.Prefix with
        | Some prefix -> parts.Add($":{prefix}")
        | None -> ()

        parts.Add message.Command

        message.Params
        |> List.iteri (fun index parameter ->
            let isLast = index = message.Params.Length - 1
            parts.Add(if isLast && parameter.Contains " " then $":{parameter}" else parameter))

        String.concat " " parts + "\r\n"
        |> Encoding.UTF8.GetBytes
